File: src/main/kotlin/antlr/Function.java
package antlr;

import antlr.KongParser.ExpressionContext;

import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.TerminalNode;

import java.util.List;
import java.util.Map;

public class Function {

  private String id;
  private List<TerminalNode> params;
  private ParseTree block;

  public Function(String id, List<TerminalNode> params, ParseTree block) {
    this.id = id;
    this.params = params;
    this.block = block;
  }

  public NodeValue invoke(List<ExpressionContext> params, Map<String, Function> functions, Scope scope) {
    if (params.size() != this.params.size()) {
      throw new RuntimeException("Illegal Function call");
    }
    scope = new Scope(scope); // create function scope
    EvalVisitor evalVisitor = new EvalVisitor(scope, functions);
    for (int i = 0; i < this.params.size(); i++) {
      NodeValue value = evalVisitor.visit(params.get(i));
      scope.assignParam(this.params.get(i).getText(), value);
    }
    NodeValue ret = NodeValue.VOID;
    try {
      evalVisitor.visit(this.block);
    } catch (ReturnValue returnValue) {
      ret = returnValue.value;
    }
    return ret;
  }
}

File: src/main/kotlin/antlr/Function.kt
package antlr

import antlr.KongParser.ExpressionContext

import org.antlr.v4.runtime.tree.ParseTree
import org.antlr.v4.runtime.tree.TerminalNode

class Function(val id: String, val params: List<TerminalNode>, val block: ParseTree) {

    fun invoke(params: List<ExpressionContext>, functions: Map<String, Function>, outerScope: Scope?): NodeValue {
        if (params.size != this.params.size) {
            throw RuntimeException("Illegal Function call")
        }
        var functionScope = Scope(outerScope) // create function functionScope
        val evalVisitor = EvalVisitor(functionScope, functions)
        for (i in this.params.indices) {
            val value = evalVisitor.visit(params[i])
            functionScope.assignParam(this.params[i].text, value)
        }
        var ret: NodeValue?
        try {
            ret = evalVisitor.visit(this.block)
        } catch (returnValue: ReturnValue) {
            ret = returnValue.value
        }

        return ret ?: NodeValue.VOID
    }
}
